libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+    segment[SEG_DECEL_END].end_vel = MAX(0.0f, segment[SEG_DECEL_END].end_vel);
+
+    // add to constant velocity segment to end at the correct position
+    const float dP = MAX(0.0f, track_length - segment[SEG_DECEL_END].end_pos);
+    const float t15 =  dP / segment[SEG_CONST].end_vel;
+    for (uint8_t i = SEG_CONST; i <= SEG_DECEL_END; i++) {
+        segment[i].end_time += t15;
+        segment[i].end_pos += dP;
+    }
+
+    // catch calculation errors
+    if (!valid()) {
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        ::printf("SCurve::set_origin_speed_max invalid path\n");
+        debug();
+#endif
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
+        init();
+        return 0.0f;
+    }
+
+    return speed;
+}
+
+// set the maximum vehicle speed at the destination
+void SCurve::set_destination_speed_max(float speed)
+{
+    // if path is zero length then start speed must be zero
+    if (num_segs != segments_max) {
+        return;
+    }
+
+    // avoid re-calculating if unnecessary
+    if (is_equal(segment[segments_max-1].end_vel, speed)) {
+        return;
+    }
+
+    const float Vm = segment[SEG_CONST].end_vel;
+    const float track_length = track.length();
+    speed = MIN(speed, Vm);
+
+    float Jm, t2, t4, t6;
+    calculate_path(jerk_time, jerk_max, speed, accel_max, Vm, track_length / 2.0f, Jm, t2, t4, t6);
+
+    uint8_t seg = SEG_CONST;
+    add_segment_const_jerk(seg, 0.0f, 0.0f);
+
+    add_segments_jerk(seg, jerk_time, -Jm, t6);
+    add_segment_const_jerk(seg, t4, 0.0f);
+    add_segments_jerk(seg, jerk_time, Jm, t2);
+
+    // remove numerical errors
+    segment[SEG_DECEL_END].end_accel = 0.0f;
+    segment[SEG_DECEL_END].end_vel = MAX(0.0f, segment[SEG_DECEL_END].end_vel);
+
+    // add to constant velocity segment to end at the correct position
+    const float dP = MAX(0.0f, track_length - segment[SEG_DECEL_END].end_pos);
+    const float t15 =  dP / segment[SEG_CONST].end_vel;
+    for (uint8_t i = SEG_CONST; i <= SEG_DECEL_END; i++) {
+        segment[i].end_time += t15;
+        segment[i].end_pos += dP;
+    }
+
+    // catch calculation errors
+    if (!valid()) {
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        ::printf("SCurve::set_destination_speed_max invalid path\n");
+        debug();
+#endif
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
+        init();
+    }
+}
+
+// move target location along path from origin to destination
+// prev_leg and next_leg are the paths before and after this path
+// wp_radius is max distance from the waypoint at the apex of the turn
+// fast_waypoint should be true if vehicle will not stop at end of this leg
+// dt is the time increment the vehicle will move along the path
+// target_pos should be set to this segment's origin and it will be updated to the current position target
+// target_vel and target_accel are updated with new targets
+// returns true if vehicle has passed the apex of the corner
```

Comment:
This is at odds with the use of the return value as `s_finished` in `AC_WPNav.cpp`

Reply:
no, not as far as I know. I think the s_finished is becoming true when the vehicle reaches the apex of the corner.